xernel/kernel/src/mem/paging.rs

```diff
@@ -293,11 +293,79 @@ impl Pagemap {
             (*pml1)[virt.p1_index()].set_unused();
         }
     }
+
+    /// Only works with 4KiB pages
+    pub fn translate(&self, virt: VirtAddr) -> Option<PhysAddr> {
+        let pml4 = self.page_table;
+
+        unsafe {
+            if !(*pml4)[virt.p4_index()].flags().contains(PageTableFlags::PRESENT) {
+                return None;
+            }
+
+            let pml3 = (*pml4)[virt.p4_index()].addr().as_u64() as *mut PageTable;
+
+            if !(*pml3)[virt.p3_index()].flags().contains(PageTableFlags::PRESENT) {
+                return None;
+            }
+
+            let pml2 = (*pml3)[virt.p3_index()].addr().as_u64() as *mut PageTable;
+
+            if !(*pml2)[virt.p2_index()].flags().contains(PageTableFlags::PRESENT) {
+                return None;
+            }
+
+            let pml1 = (*pml2)[virt.p2_index()].addr().as_u64() as *mut PageTable;
+
+            if !(*pml1)[virt.p1_index()].flags().contains(PageTableFlags::PRESENT) {
+                return None;
+            }
+
+            Some((*pml1)[virt.p1_index()].addr() + u64::from(virt.page_offset()))
+        }
+    }
+
+    fn deallocate_pt(pt: *mut PageTable, level: u8) {
+        let mut frame_allocator = FRAME_ALLOCATOR.lock();
+        if level == 1 {
+            for i in 0..128 {
+                unsafe {
+                    if (*pt)[i].flags().contains(PageTableFlags::PRESENT) {
+                        let pt = ((*pt)[i].addr().as_u64() + *HIGHER_HALF_OFFSET) as *mut PageTable;
+
+                        Self::deallocate_pt(pt, level + 1);
```

Comment:
If level 1 is highest (cause we only go from 0..128), then why increase level here?
Provide explanation (since normally level 4 is highest)